pkg/ruler/rules/objectclient/rule_store.go

```diff
@@ -74,25 +75,56 @@ func (o *RuleStore) ListAllRuleGroups(ctx context.Context) (map[string]rules.Rul
 		return nil, err
 	}
 
+	mtx := sync.Mutex{}
 	userGroupMap := map[string]rules.RuleGroupList{}
-	for _, obj := range ruleGroupObjects {
-
-		user := decomposeRuleObjectKey(obj.Key)
-		if user == "" {
-			continue
-		}
-
-		rg, err := o.getRuleGroup(ctx, obj.Key)
-		if err != nil {
-			return nil, err
-		}
+	ch := make(chan string)
+	errCh := make(chan error)
```

Comment:
Can we use [errgroup](https://godoc.org/golang.org/x/sync/errgroup) here? That should remove need for `WaitGroup` and error channel, and will stop all goroutines when context is cancelled.

(Unbuffered channel here will cause leak of gouroutines if multiple workers try to send error)